packages/plugin-ext/src/main/browser/view/tree-view-widget.tsx

```diff
@@ -677,21 +677,21 @@ export class TreeViewWidget extends TreeViewWelcomeWidget {
                         }
                     }
                 }
-                if (items.length > 0 || event.dataTransfer.types.indexOf(this.treeDragType) >= 0) {
+                if (items.length > 0 || event.dataTransfer.types.includes(this.treeDragType)) {
                     event.preventDefault();
                     event.stopPropagation();
-                    const filesCopy = [...files];
-                    this.model.proxy?.$drop(this.id, node?.id, items, CancellationToken.None).then(() => {
-                        for (const file of filesCopy) {
+                    this.model.proxy?.$drop(this.id, node?.id, items, CancellationToken.None).finally(() => {
+                        for (const file of files) {
                             this.dndFileContentStore.removeFile(file);
                         }
                     });
                     files = [];
                 }
-            } finally {
+            } catch (e) {
                 for (const file of files) {
                     this.dndFileContentStore.removeFile(file);
                 }
+                throw e;
```

Comment:
The new logic looks more solid. Now the only slightly (and only slightly) troubling bit is that the connection between `items` and `files` is implicit. According to the current logic, there will only be `files` if there are `items`, so we always clean `files` up if there are any, because we always clean them up if there are `items`. But that connection isn't expressed in the code, so it would be possible in the future to break it (by adding a stricter filter to 'itemness' for example), and then we might not hit either 685 or 692. But that's a strictly hypothetical issue.

Reply:
I don't really understand what you mean here: we need to forget every file we remember after the drop operation is over, either by an sync exception or when the `$drop` promise finishes (resolve or reject). There never will be any file in the `dndFileContentStore` when no drop operation is going on.

Reply:
My only point is that the connection between the `items` array that determines whether we get into the `$drop` branch or not and the `files` array that we clear if we don't get into that branch are only implicitly connected. In the current logic, everything should be alright, but if there were every a disconnect between the `items` and the `files`, then it would be possible to add things to the store and not clear them.

Reply:
I just don't see how that could happen: we add the ids of the files we put into the store into the `files` array immediately after we put them into the store (#6623), so every remembered file will be in the `files` array and the files array is what we use to release the files again. The items just never figure into the equation. If you're still worried, you'll have to explain it to me like im five ;-)

Reply:
You're right about the current logic - it's solid. But because `items` is one thing and `files` is another, it would be possible for their current synchronicity to be broken in the future. The solution I can see would be not to have a separate `files` array, but add the file handle to the item itself - just here, not as part of the interface - and then use the handle stored on the item when clearing the cache. That would make explicit the 1:1 mapping of items and files. But I'm not really worried about it.

Reply:
But there isn't a 1:1 mapping: some items have files, and others don't, so it would be a 1:[0, 1]. I don't think the spec guarantees that all items being dropped are either all files or all not files.